kong/cache_warmup.lua

```diff
@@ -0,0 +1,140 @@
+local utils = require "kong.tools.utils"
+
+
+local cache_warmup = {}
+
+
+local tostring = tostring
+local ipairs = ipairs
+local math = math
+local kong = kong
+local ngx = ngx
+
+
+function cache_warmup._mock_kong(mock_kong)
+  kong = mock_kong
+end
+
+
+local function warmup_dns(premature, hosts, count)
+  if premature then
+    return
+  end
+
+  ngx.log(ngx.NOTICE, "warming up DNS entries ...")
+
+  local start = ngx.now()
+
+  for i = 1, count do
+    kong.dns.toip(hosts[i])
+  end
+
+  local elapsed = math.floor((ngx.now() - start) * 1000)
+
+  ngx.log(ngx.NOTICE, "finished warming up DNS entries",
+                      "' into the cache (in ", tostring(elapsed), "ms)")
+end
+
+
+local function fail_cb()
+  error("this should never be called as L2 should already be warmed")
+end
+
+
+local function cache_warmup_single_entity(dao)
+  local entity_name = dao.schema.name
+
+  ngx.log(ngx.NOTICE, "Preloading '", entity_name, "' into the cache ...")
+
+  local start = ngx.now()
+
+  local hosts, names, host_count
+  if entity_name == "services" then
+    hosts = {}
+    names = {}
+    host_count = 0
+  end
+
+  for entity, err in dao:each(1000) do
+    if err then
+      return nil, err
+    end
+
+    if entity_name == "services" then
+      if utils.hostname_type(entity.host) == "name"
+         and names[entity.host] == nil then
+        host_count = host_count + 1
+        hosts[host_count] = entity.host
+        names[entity.host] = true
+      end
+    end
+
+    local cache_key = dao:cache_key(entity)
+
+    local ok, err = kong.cache:safe_set(cache_key, entity)
+    if not ok then
+      return nil, err
+    end
+
+    -- NOTE: this is just for warming up L1
+    ok, err = kong.cache:get(cache_key, nil, fail_cb)
+    if not ok then
+      return nil, err
+    end
+  end
+
+  if entity_name == "services" and host_count > 0 then
+    ngx.timer.at(0, warmup_dns, hosts, host_count)
+  end
+
+  local elapsed = math.floor((ngx.now() - start) * 1000)
+
+  ngx.log(ngx.NOTICE, "finished preloading '", entity_name,
+                      "' into the cache (in ", tostring(elapsed), "ms)")
+  return true
+end
+
+
+-- Loads entities from the database into the cache, for rapid subsequent
+-- access. This function is intented to be used during worker initialization.
+function cache_warmup.execute(entities)
+  -- kong.db and kong.cache might not be active while running tests
+  if not kong.db or not kong.cache then
+    return true
+  end
+
+  for _, entity_name in ipairs(entities) do
+    if entity_name == "routes" then
```

Comment:
I'd rather not skip Routes silently. Either of these options would would do it for me:
* We allow the user to fill the cache with routes (maybe they do something with routes in a custom plugin?)
* We forbid them explicitly (by making them an invalid option in `conf.db_cache_warmup_entities`)

If none of these options is possible, we should at least put a `WARN` in the logs indicating that routes are never prewarmed.

Reply:
Because it is an explicit setting, and not enabled by default, then I would go with:
```
We allow the user to fill the cache with routes (maybe they do something with routes in a custom plugin?)
```

Reply:
The current behavior was added at @thibaultcha's request. @thibaultcha, what do you say?

> If none of these options is possible, we should at least put a WARN in the logs indicating that routes are never prewarmed.

Adding `routes` to the list in the current behavior is harmless, so `warn` is IMO overreacting, but I wouldn't mind adding a `notice` level log.

Reply:
In case we go down that route (no pun intended!), I already added the `kong.log.notice` log.